rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/custom_participant_info.hpp

```diff
@@ -71,6 +88,54 @@ typedef struct CustomParticipantInfo
   // with the default configuration.
   bool leave_middleware_default_qos;
   publishing_mode_t publishing_mode;
+
+  eprosima::fastdds::dds::Topic * find_or_create_topic(
+    const std::string & topic_name,
+    const std::string & type_name,
+    const eprosima::fastdds::dds::TopicQos & topic_qos)
+  {
+    eprosima::fastdds::dds::Topic * topic = nullptr;
+
+    std::lock_guard<std::mutex> lck(topic_name_to_topic_mutex_);
+    std::map<std::string,
+      std::unique_ptr<UseCountTopic>>::const_iterator it = topic_name_to_topic_.find(topic_name);
+    if (it == topic_name_to_topic_.end()) {
+      // Not already in the map, we need to add it
+      topic = participant_->create_topic(topic_name, type_name, topic_qos);
+
+      auto uct = std::make_unique<UseCountTopic>();
+      uct->topic = topic;
+      uct->use_count = 1;
+
+      topic_name_to_topic_[topic_name] = std::move(uct);
+    } else {
+      // Already in the map, just increase the use count
+      it->second->use_count++;
+      topic = it->second->topic;
+    }
+
+    return topic;
+  }
+
+  void delete_topic(const eprosima::fastdds::dds::Topic * topic)
+  {
+    if (topic == nullptr) {
+      return;
+    }
+
+    std::lock_guard<std::mutex> lck(topic_name_to_topic_mutex_);
+    std::map<std::string, std::unique_ptr<UseCountTopic>>::const_iterator it =
+      topic_name_to_topic_.find(topic->get_name());
+
+    if (it != topic_name_to_topic_.end()) {
```

Comment:
Should we emit a warning if the topic was not found?

Reply:
Yeah, good point.  Added in 06ab8a9d1b305ea359560de2182106a53504d690.